client/rpc/src/chain/chain_full.rs

```diff
@@ -68,27 +72,104 @@ where
 	}
 
 	fn subscribe_all_heads(&self, sink: SubscriptionSink) -> Result<(), Error> {
-		let client = self.client.clone();
-		let executor = self.executor.clone();
-
-		let fut = helpers::subscribe_headers(client, sink, "chain_subscribeAllHeads");
-		executor.spawn_obj(Box::pin(fut).into()).map_err(|e| Error::Client(Box::new(e)))
+		subscribe_headers(
+			&self.client,
+			self.executor.clone(),
+			"chain_subscribeAllHeads",
+			sink,
+			|| self.client().info().best_hash,
+			|| {
+				self.client()
+					.import_notification_stream()
+					.map(|notification| notification.header)
+			},
+		)
 	}
 
 	fn subscribe_new_heads(&self, sink: SubscriptionSink) -> Result<(), Error> {
-		let client = self.client.clone();
-		let executor = self.executor.clone();
-
-		let fut = helpers::subscribe_headers(client, sink, "chain_subscribeNewHeads");
-		executor.spawn_obj(Box::pin(fut).into()).map_err(|e| Error::Client(Box::new(e)))
+		subscribe_headers(
+			&self.client,
+			self.executor.clone(),
+			"chain_subscribeNewHeads",
+			sink,
+			|| self.client().info().best_hash,
+			|| {
+				self.client()
+					.import_notification_stream()
+					.filter(|notification| future::ready(notification.is_new_best))
+					.map(|notification| notification.header)
+			},
+		)
 	}
 
 	fn subscribe_finalized_heads(&self, sink: SubscriptionSink) -> Result<(), Error> {
-		let client = self.client.clone();
-		let executor = self.executor.clone();
+		subscribe_headers(
+			&self.client,
+			self.executor.clone(),
+			"chain_subscribeFinalizedHeads",
+			sink,
+			|| self.client().info().finalized_hash,
+			|| {
+				self.client()
+					.finality_notification_stream()
+					.map(|notification| notification.header)
+			},
+		)
+	}
+}
+
+/// Subscribe to new headers.
+fn subscribe_headers<Block, Client, F, G, S>(
+	client: &Arc<Client>,
+	executor: SubscriptionTaskExecutor,
+	method: &'static str,
+	mut sink: SubscriptionSink,
+	best_block_hash: G,
+	stream: F,
+) -> Result<(), Error>
+where
+	Block: BlockT + 'static,
+	Block::Header: Unpin,
+	Client: HeaderBackend<Block> + 'static,
+	F: FnOnce() -> S,
+	G: FnOnce() -> Block::Hash,
+	S: Stream<Item = Block::Header> + Send + 'static,
+{
+		// send current head right at the start.
+		let maybe_header = client
+			.header(BlockId::Hash(best_block_hash()))
+			.map_err(client_err)
+			.and_then(|header| {
+				header.ok_or_else(|| Error::Other("Best header missing.".to_string()))
+			})
+			.map_err(|e| {
+				log::warn!("Best header error {:?}", e);
+				e
+			})
+			.ok();
+
+		// send further subscriptions
+		let stream = stream();
+
+		// NOTE: by the time we set up the stream there might be a new best block and so there is a risk
+		// that the stream has a hole in it. The alternative would be to look up the best block *after*
+		// we set up the stream and chain it to the stream. Consuming code would need to handle
+		// duplicates at the beginning of the stream though.
+		let fut = async move {
+			stream::iter(maybe_header)
```

Comment:
This will only send the result if `maybe_header.is_some()` it should be infallible to get the best_block AFAIU....

More similar to how `substrate master` behaves FWIW.

Reply:
See https://github.com/paritytech/substrate/blob/master/client/rpc/src/chain/mod.rs#L301-#L322

Reply:
> it should be infallible to get the best_block AFAIU....

Well, it's a db lookup right? So sure, if it fails the node has bigger problems, but it's not infallible.

Not sure if you suggest we should get rid of the comment about "holes" in the stream, but I don't think this new code changes that. If there's a new block between when we lookup the best block (i.e. resolve `maybe_header`) and the time we read from the stream, the block we sent as "best" might not be immediately precedent to the next one we send. I don't think this is a catastrophic flaw though, more like something to be aware of.

Reply:
> Well, it's a db lookup right? So sure, if it fails the node has bigger problems, but it's not infallible.

Yeah, you right I was under impression that it was cached.

My point was that before the error was sent over the channel now it's just logged and filtered out via `stream:iter(None)`.

Reply:
Thus, the comment is still "accurate"